File: algorithms/127_word_ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordDict) {
        wordDict.emplace(endWord);
        deque<string> temp;
        temp.push_back(beginWord);
        int length=1;
        while(!temp.empty()) {
            int que_size=temp.size();
            length++;
            for (int i=0;i<que_size;i++) {
                string pop_str=temp.front();
                stack<string> erase_stack;
                temp.pop_front();
                for (auto it=wordDict.begin();it!=wordDict.end();it++) {
                    if (matchOne(pop_str,*it)) {
                        if (*it==endWord) return length;;
                        temp.push_back(*it);
                        erase_stack.push(*it);}
                    }
                while(!erase_stack.empty()) {
                    wordDict.erase(erase_stack.top());
                    erase_stack.pop();
                }
            }
        }
        return 0; 
    }
    
    bool matchOne(const string& word1, const string& word2) {
        if (word1.size()!= word2.size()) return false;
        int count=0;
        for (int i=0;i<word1.size();i++) {
            if (word1[i]!=word2[i]) count++;
            if (count>1) break;}
        return (count==1);
        }
};

File: algorithms/python3/127_word_ladder.py
from collections import defaultdict
from typing import List
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if not beginWord or not endWord or not wordList or not endWord in wordList:
            return 0
        word_dict=defaultdict(list)
        l=len(beginWord)
        for word in wordList:
            for i in range(l):
                word_dict[word[:i]+'*'+word[i+1:]].append(word)
        queue=[(beginWord,1)]
        visited={beginWord:True}
        while queue:
            curword,level=queue.pop(0)
            for i in range(l):
                inter_word=curword[:i]+'*'+curword[i+1:]
                for word in word_dict[inter_word]:
                    if word==endWord:
                        return level+1
                    if word not in visited:
                        visited[word]=True
                        queue.append((word,level+1))
            word_dict[inter_word]=[]
        return 0


